File: Supervisor/Supervisor.R
library(tidyverse)
library(car)
library(GGally)
library(MASS)

# Read data
supervisor <- read.table(file = "https://mheaton.byu.edu/Courses/Stat330/InClassAnalyses/4%20-%20SupervisorPerformance/Data/Supervisor.txt", sep = "", header = TRUE)

# Build Matrix scatter plots
ggpairs(supervisor)

# Fitting model
mlr <- lm(Rating ~ ., data = supervisor)
summary(mlr)

# Std.Residual histogram
std <- stdres(mlr)
hist(std)

# Added variable plots
avPlots(mlr)

# Residual plots
plot(mlr$residuals ~ mlr$fitted.values)
abline(h=0)

# Try prediction
predict.lm (mlr,newdata=data.frame(Complaints = 65, Privileges = 51.5, Learn = 56.5, Raises = 63.5, Critical = 77.5, Advance = 41), interval="prediction", level = 0.95)

# Cross Validation
n.cv <- 250
bias <- rep(NA,n.cv)
rpmse <- rep(NA,n.cv)
cvg <- rep(NA,n.cv)
wid <- rep(NA,n.cv)
n.test <- nrow(supervisor)/10

for(i in 1:n.cv){
  #split into test and training set
  test.obs <- sample(1:nrow(supervisor),n.test)
  test.set <- supervisor[test.obs,]
  train.set <- supervisor[-test.obs,]
  
  # fit a lm using training data only
  train.lm <- lm(Rating ~ ., data=train.set)
  
  # Prediction and prediction intervals
  test.pred <- predict.lm(train.lm,newdata = test.set,interval="prediction",level = 0.95)  
  
  # calculate results
  bias[i] <- mean(test.pred[,1] - test.set$Rating)
  rpmse[i] <- sqrt(mean((test.pred[,1] - test.set$Rating)^2))
  cvg[i] <- mean(test.pred[,2] < test.set$Rating & test.pred[,3] > test.set$Rating)
  wid[i] <- mean(test.pred[,3]-test.pred[,2])  
}
mean(bias)
mean(rpmse)
mean(cvg)
mean(wid)

File: Supervisor/Supervisor.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.formula.api as smf
import statsmodels.stats.api as sm
import statsmodels.graphics.api as smg

# Reading in Data
supervisor = pd.read_table("/Users/ldskimjg/Desktop/330/4.Supervisor/Supervisor.txt", sep = " ")
print(supervisor)
supervisor.shape

# pair plot
sns.pairplot(supervisor)
supervisor.corr()

# Fitting model
explan_vars = " + ".join(supervisor.columns[1:])
explan_vars
mylm = smf.ols("Rating ~" + explan_vars, data = supervisor).fit()
mylm.summary()
mylm.mse_resid
mylm.params
mylm.rsquared

# Linearity - added variable plot
smg.plot_partregress_grid(mylm)

# BP test & plot of residuals vs. fitted values
sns.regplot(x = mylm.fittedvalues, y = mylm.resid, color = 'blue', ci = None)
sm.het_breuschpagan(mylm.resid_pearson, mylm.model.exog)

# Normality
sns.distplot(mylm.resid_pearson, kde = False)
mylm.summary()

# Cross Validation
ncv = 250
bias = np.repeat(np.NaN,ncv)
rpmse = np.repeat(np.NaN,ncv)
wid = np.repeat(np.NaN,ncv)
ntest = round(supervisor.shape[0]/10)

for cv in range(ncv) :
    # Choose which obs to put
    testobs = np.random.choice(supervisor.shape[0], ntest)
    
    # Split data
    testset = supervisor.iloc[testobs,:]
    trainset = supervisor.drop(testobs)
    
    # Using training data to fit a model
    trainlm = smf.ols("Rating ~" + explan_vars, data = trainset).fit()
    
    # Predict test set
    testpreds = trainlm.get_prediction(testset)
    
    # Calculate bias
    bias[cv] = np.mean(testpreds.predicted_mean - testset['Rating'])
    
    # Calculate RPMSE
    rpmse[cv] = np.sqrt(np.mean(np.square(testpreds.predicted_mean-testset['Rating'])))
    
    # Calculate Coverage
    pis = testpreds.conf_int(obs=True, alpha = 0.05)
    conditions_array = [pis[:,0] <= testset['Rating'].values, pis[:,1] >= testset['Rating'].values]
    cond_met = np.all(conditions_array, axis = 0)
    
    # Calculate Width
    wid[cv] = np.mean(pis[:,1] - pis[:,0])

np.mean(bias)
np.mean(rpmse)
np.mean(cond_met)
np.mean(wid)

# 95% confidence intervals
mylm.conf_int(alpha=0.05)

# 95% prediction intervals

predframe = pd.DataFrame(dict(Complaints = [65], Privileges = [51.5], Learn = [56.5], Raises = [63.5], Critical = [77.5], Advance = [41]))

preds = mylm.get_prediction(predframe)

preds.summary_frame()
preds.predicted_mean
preds.conf_int(obs = True, alpha = 0.05)
